server/src/main/java/org/opensearch/gateway/BaseShardCache.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    /**
+     * Finds all the nodes that need to be fetched. Those are nodes that have no
+     * data, and are not in fetch mode.
+     */
+    List<String> findNodesToFetch() {
+        List<String> nodesToFetch = new ArrayList<>();
+        for (BaseNodeEntry nodeEntry : getCache().values()) {
+            if (nodeEntry.hasData() == false && nodeEntry.isFetching() == false) {
+                nodesToFetch.add(nodeEntry.getNodeId());
+            }
+        }
+        return nodesToFetch;
+    }
+
+    /**
+     * Are there any nodes that are fetching data?
+     */
+    boolean hasAnyNodeFetching() {
+        for (BaseNodeEntry nodeEntry : getCache().values()) {
+            if (nodeEntry.isFetching()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get the data from cache, ignore the failed entries. Use getData functional interface to get the data, as
+     * different implementations may have different ways to populate the data from cache.
+     *
+     * @param nodes       Discovery nodes for which we need to return the cache data.
+     * @param failedNodes return failedNodes with the nodes where fetch has failed.
+     * @return Map of cache data for every DiscoveryNode.
+     */
+    Map<DiscoveryNode, K> populateCache(DiscoveryNodes nodes, Set<String> failedNodes) {
+        Map<DiscoveryNode, K> fetchData = new HashMap<>();
+        for (Iterator<? extends Map.Entry<String, ? extends BaseNodeEntry>> it = getCache().entrySet().iterator(); it.hasNext();) {
+            Map.Entry<String, BaseNodeEntry> entry = (Map.Entry<String, BaseNodeEntry>) it.next();
+            String nodeId = entry.getKey();
+            BaseNodeEntry nodeEntry = entry.getValue();
+
+            DiscoveryNode node = nodes.get(nodeId);
+            if (node != null) {
+                if (nodeEntry.isFailed()) {
+                    // if its failed, remove it from the list of nodes, so if this run doesn't work
+                    // we try again next round to fetch it again
+                    it.remove();
+                    failedNodes.add(nodeEntry.getNodeId());
+                } else {
+                    K nodeResponse = getData(node);
+                    if (nodeResponse != null) {
+                        fetchData.put(node, nodeResponse);
+                    }
+                }
+            }
+        }
+        return fetchData;
+    }
+
+    void processResponses(List<K> responses, long fetchingRound) {
+        for (K response : responses) {
+            BaseNodeEntry nodeEntry = getCache().get(response.getNode().getId());
+            if (nodeEntry != null) {
+                if (validateNodeResponse(nodeEntry, fetchingRound)) {
+                    // if the entry is there, for the right fetching round and not marked as failed already, process it
+                    logger.trace("{} marking {} as done for [{}], result is [{}]", logKey, nodeEntry.getNodeId(), type, response);
```

Comment:
logKey can be `ShardId` or `BatchId` based on which type of cache is using these methods. 
Do you think providing index in `getLogger` would make sense ? 

Right now this `logKey` is used  to debug the issues for a specific shardId or batchId. For batch scenario, we can't pass a single index because one batch can have shards of different indices also. 

Please let me know if this is okay or some other way is there for this use case.

Reply:
```
 public static Logger getLogger(Logger parentLogger, String s) {
        Logger inner = LogManager.getLogger(parentLogger.getName() + s);
        if (parentLogger instanceof PrefixLogger) {
            return new PrefixLogger(inner, ((PrefixLogger) parentLogger).prefix());
        }
        return inner;
    }
```

This method allows us to include the prefix in the logger name itself. This can be used to avoid adding logKey each time. 

Another method of this class is
```
public static Logger getLogger(Class<?> clazz, String... prefixes) {
        return new PrefixLogger(LogManager.getLogger(clazz), formatPrefix(prefixes));
    }
```

requires a class in the input which can't be provided in current implementation as we're already receiving a logger instance from caller class. 

Planning to use the logger like - `Loggers.getLogger(logger, "_" + logKey)` . This way logs can be grepped using logKey as well as **BaseShardCache** doesn't need to remember logKey.

Let me know if this looks fine ?